sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +106,103 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
                 ProcessNext(message, pipeline);
             }
         }
+
+        private class AccessTokenCache
+        {
+            private readonly object _syncObj = new object();
+            private readonly TimeSpan _tokenRefreshOffset;
+
+            private TokenState _tokenState;
+            private string? _headerValue;
+            private TaskCompletionSource<string>? _pendingTcs;
+            private DateTimeOffset _refreshOn;
+            private DateTimeOffset _expiresOn;
+
+            public AccessTokenCache(TimeSpan tokenRefreshOffset)
+            {
+                _tokenRefreshOffset = tokenRefreshOffset;
+            }
+
+            public string? SaveToken(AccessToken token, Exception? exception)
+            {
+                string? headerValue;
+                TaskCompletionSource<string>? pendingTcs;
+
+                lock (_syncObj)
+                {
+                    if (!string.IsNullOrEmpty(token.Token))
+                    {
+                        _headerValue = "Bearer " + token.Token;
+                        _refreshOn = token.ExpiresOn - _tokenRefreshOffset;
+                        _expiresOn = token.ExpiresOn;
+                        _tokenState = TokenState.Valid;
+                    }
+                    else
+                    {
+                        _headerValue = default;
+                        _refreshOn = default;
+                        _expiresOn = default;
+                        _tokenState = TokenState.Invalid;
+                    }
+
+                    pendingTcs = _pendingTcs;
+                    headerValue = _headerValue;
+                    _pendingTcs = null;
+                }
+
+                if (pendingTcs != null)
+                {
+                    if (headerValue != null)
+                    {
+                        pendingTcs.SetResult(headerValue);
+                    }
+                    else
+                    {
+                        pendingTcs.SetException(exception ?? new InvalidOperationException());
+                    }
+                }
+
+                return headerValue;
+            }
+
+            public (string? headerValue, Task<string>? pendingTask) UpdateTokenState()
```

Comment:
Why don't you return a ValueTask<string> here?

Reply:
You mean to use the `ManualResetValueTaskSourceCore` instead of `TaskCompletionSource`?

Reply:
No, don't return the tuple and handle actual value retrieval inside AccessTokenCache

Reply:
Done